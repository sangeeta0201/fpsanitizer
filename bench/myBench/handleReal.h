#include <iostream>
#include <map>
#include <math.h>
#include <gmp.h>
#include <mpfr.h>

#define PRECISION 200

std::map<size_t, struct Real*> shadowMap;
struct Real{
  mpfr_t mpfr_val;
};

double getDouble(Real *real);
unsigned long ulpd(double x, double y);
void handleOp(size_t opCode, mpfr_t *res, mpfr_t *op1, mpfr_t *op2);
double updateError(Real *realVal, double computedVal);
void printError(size_t result_A, double y);

extern "C" void* handleOp_1(size_t opCode, void *op1, void *op2);
extern "C" void* handleOp_2(size_t opCode, float op1, void *op2);
extern "C" void* handleOp_3(size_t opCode, void *op1, float op2);
extern "C" void* handleOp_4(size_t opCode, float op1, float op2);
/*
When we compute any floating point operation on real, we save the result in shadow map with its value 
and return the address of this saved value to LLVM Pass. LLVM pass creates the mapping of real result and 
floating result. Wherever floating result is used, we use real result to compute the parallel operation in real.
Whenever store is called to save the value of temp into variable, we call setRealTemp. This function save the real operation 
result into variable.  
we create a ma
For every store, we call setRealConstant if  store is storing constant to some variable. 
We create shadow space for this variable, which is address of varaible and the real constant 
of this variable.
It 
*/
extern "C" void setRealTemp(void *toAddr, void *fromAddr);

/*
For every store, we call setRealConstant if  store is storing constant to some variable. 
We create shadow space for this variable, which is address of varaible and the real constant 
of this variable.
It 
*/
extern "C" void setRealConstant_float(void *Addr, float value);
extern "C" void setRealConstant_double(void *Addr, double value);
