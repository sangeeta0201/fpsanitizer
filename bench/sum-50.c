#include <stdio.h>

int main() {
  volatile double x;
  for (x = 0.0; x < 2.0; x += 0.2){
  	printf("%.20g\n", x);
  }
}
