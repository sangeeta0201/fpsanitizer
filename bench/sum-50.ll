; ModuleID = 'sum-50.bc'
source_filename = "sum-50.c"
target datalayout = "e-m:e-i64:64-f80:128-n8:16:32:64-S128"
target triple = "x86_64-unknown-linux-gnu"

@.str = private dso_local unnamed_addr constant [7 x i8] c"%.20g\0A\00", align 1

; Function Attrs: nounwind uwtable
define dso_local i32 @main() local_unnamed_addr #0 {
entry:
  %x = alloca double, align 8
  %x.0.x.0..sroa_cast = bitcast double* %x to i8*
  call void @llvm.lifetime.start.p0i8(i64 8, i8* nonnull %x.0.x.0..sroa_cast)
  store volatile double 0.000000e+00, double* %x, align 8, !tbaa !2
  %x.0.x.0.4 = load volatile double, double* %x, align 8, !tbaa !2
  %cmp5 = fcmp olt double %x.0.x.0.4, 2.000000e+00
  %x.0.x.0.16 = load volatile double, double* %x, align 8, !tbaa !2
  br i1 %cmp5, label %for.inc, label %for.end

for.inc:                                          ; preds = %entry, %for.inc
  %x.0.x.0.17 = phi double [ %x.0.x.0.1, %for.inc ], [ %x.0.x.0.16, %entry ]
  %add = fadd double %x.0.x.0.17, 2.000000e-01
  store volatile double %add, double* %x, align 8, !tbaa !2
  %x.0.x.0. = load volatile double, double* %x, align 8, !tbaa !2
  %cmp = fcmp olt double %x.0.x.0., 2.000000e+00
  %x.0.x.0.1 = load volatile double, double* %x, align 8, !tbaa !2
  br i1 %cmp, label %for.inc, label %for.end

for.end:                                          ; preds = %for.inc, %entry
  %x.0.x.0.1.lcssa = phi double [ %x.0.x.0.16, %entry ], [ %x.0.x.0.1, %for.inc ]
  %call = tail call i32 (i8*, ...) @printf(i8* getelementptr inbounds ([7 x i8], [7 x i8]* @.str, i64 0, i64 0), double %x.0.x.0.1.lcssa)
  call void @llvm.lifetime.end.p0i8(i64 8, i8* nonnull %x.0.x.0..sroa_cast)
  ret i32 0
}

; Function Attrs: argmemonly nounwind
declare void @llvm.lifetime.start.p0i8(i64, i8* nocapture) #1

; Function Attrs: nounwind
declare dso_local i32 @printf(i8* nocapture readonly, ...) local_unnamed_addr #2

; Function Attrs: argmemonly nounwind
declare void @llvm.lifetime.end.p0i8(i64, i8* nocapture) #1

attributes #0 = { nounwind uwtable "correctly-rounded-divide-sqrt-fp-math"="false" "disable-tail-calls"="false" "less-precise-fpmad"="false" "no-frame-pointer-elim"="false" "no-infs-fp-math"="false" "no-jump-tables"="false" "no-nans-fp-math"="false" "no-signed-zeros-fp-math"="false" "no-trapping-math"="false" "stack-protector-buffer-size"="8" "target-cpu"="x86-64" "target-features"="+fxsr,+mmx,+sse,+sse2,+x87" "unsafe-fp-math"="false" "use-soft-float"="false" }
attributes #1 = { argmemonly nounwind }
attributes #2 = { nounwind "correctly-rounded-divide-sqrt-fp-math"="false" "disable-tail-calls"="false" "less-precise-fpmad"="false" "no-frame-pointer-elim"="false" "no-infs-fp-math"="false" "no-nans-fp-math"="false" "no-signed-zeros-fp-math"="false" "no-trapping-math"="false" "stack-protector-buffer-size"="8" "target-cpu"="x86-64" "target-features"="+fxsr,+mmx,+sse,+sse2,+x87" "unsafe-fp-math"="false" "use-soft-float"="false" }

!llvm.module.flags = !{!0}
!llvm.ident = !{!1}

!0 = !{i32 1, !"wchar_size", i32 4}
!1 = !{!"clang version 7.0.0 (trunk 336308)"}
!2 = !{!3, !3, i64 0}
!3 = !{!"double", !4, i64 0}
!4 = !{!"omnipotent char", !5, i64 0}
!5 = !{!"Simple C/C++ TBAA"}
