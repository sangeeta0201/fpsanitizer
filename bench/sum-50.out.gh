Result @ sum-50.c:6 in main (addr 40057D)
2.000000 bits average error
2.000000 bits max error
Aggregated over 1 instances
Influenced by erroneous expression:

No influences found!


compare @ sum-50.c:5 in main (addr 400554)
1% incorrect
1 incorrect values
52 total instances
Influenced by erroneous expressions:

No influences found!

