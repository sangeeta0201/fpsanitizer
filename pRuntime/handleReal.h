#ifndef REAL_HEADER_H
#define REAL_HEADER_H 
#include <iostream>
#include <map>
#include <queue>
#include <math.h>
#include <gmp.h>
#include <mpfr.h>
#include <vector>
#include <stack>
#include <list>
#include <assert.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <sys/mman.h>
#include "tbb/concurrent_queue.h"
#include "tbb/concurrent_hash_map.h"
#define PRECISION 1000
#define MMAP_FLAGS (MAP_PRIVATE| MAP_ANONYMOUS| MAP_NORESERVE)
#define MAX_STACK_SIZE 1000000000
#define MAX_SIZE 100000000
#define INSSIZE 1000000
#define debug 0
#define debugCR 0
//#define MPFRINIT 100000000 //for spec
#define MPFRINIT 100

const size_t SS_PRIMARY_TABLE_ENTRIES = ((size_t) 4194304);//2^22
const size_t SS_SEC_TABLE_ENTRIES = ((size_t) 64*(size_t) 1024 * (size_t) 1024); // 2^26
//const size_t SS_SEC_TABLE_ENTRIES = ((size_t)4 *(size_t) 1024 * (size_t) 1024); // 2^22

FILE *pFile = fopen ("error.log","w");
FILE *eFile = fopen ("branch.log","w");
FILE *lbmRef = fopen ("ref.log","w");

struct ErrorAggregate {
  double max_error;
  double total_error;
  long long int num_evals;
};
struct BrError {
  bool corrRes;
  bool incorrRes;
	size_t lineNo;
  long long int num_evals;
};

struct Real{
  int initFlag;
  int initMPFRFlag;
  mpfr_t mpfr_val;
};

Real *shadowStack;
Real **shadowMap;

struct Compute{
	size_t opCode;
	size_t op1Addr;
	size_t op2Addr;
	double op1d;
	double op2d;
	float op1f;
	float op2f;
	double computedResd;
	float computedResf;
	size_t insIndex;
	size_t newRegIdx;
	size_t cmd;
	size_t fcmpFlag;
	size_t lineNo;
	bool fcmpRes;
};

bool consumerFlag = false;
size_t funcCount = 0;
bool initFlag = false;
int frameIdx = 0;
size_t newRegIdx = 0;
bool recurFlag = false;
size_t mpfrClear = 0;
size_t mpfrInit = 0;
size_t mpfrClearMap = 0;
size_t mpfrInitMap = 0;
int varCount = 0;
int opCount = 0;
size_t count = 0;
size_t count1 = 0;
size_t count2 = 0;
size_t count3 = 0;
size_t count4 = 0;
size_t count5 = 0;
size_t count6 = 0;

tbb::concurrent_queue<struct Compute*> worker;
tbb::concurrent_queue<struct Compute*> ready1;
tbb::concurrent_queue<struct Compute*> ready2;
tbb::concurrent_queue<struct Compute*> ready3;
tbb::concurrent_queue<struct Compute*> ready4;
tbb::concurrent_queue<struct Compute*> ready5;
tbb::concurrent_queue<struct Compute*> ready6;
std::map<size_t, struct ErrorAggregate*>errorMap;
std::map<size_t, struct BrError*>errBrMap;
//this will link ins index to index of result in shadow mem
//std::list<struct MyShadow*> varTrack;
struct MyShadow *varTrack;
//size_t (*insMap)[100];
size_t *insMap;
size_t *frameCur;
size_t *argCount;
size_t *slotIdx;
size_t *returnIdx;
void* consumer(void *ptr);
void* consumer1(void *ptr);
void* consumer2(void *ptr);
void* consumer3(void *ptr);
void* consumer4(void *ptr);
void* consumer5(void *ptr);
void* consumer6(void *ptr);
void computeReal();

extern "C" void* __get_real_fun_arg(size_t index);
extern "C" size_t __get_addr(void *Addr);
extern "C"  void __init();
extern "C"  void __finish();
extern "C"  void printToFile(void* op1Addr, void* op2Addr, void* op3Addr);
unsigned long ulpd(double x, double y);
unsigned long ulpf(float x, float y);
void handleOp(size_t opCode, mpfr_t *res, mpfr_t *op1, mpfr_t *op2);
void initializeBrError(BrError *err);
double updateError(mpfr_t mpfr_val, double computedVal, size_t insIndex);
double updateErrorF(mpfr_t mpfr_val, float computedVal, size_t insIndex);
void updateBranchError(bool realRes, bool computedRes, size_t insIndex, size_t lineNo);
void initializeErrorAggregate(ErrorAggregate *eagg);
void printReal(mpfr_t mpfr_val);
void print_trace (void);
void addFunArg(size_t argNo, size_t argAddrInt, double op);
void funcInit();
void funcExit(size_t returnIndex);
void handleFloatToSInt(size_t op1Int, long val, size_t lineNo);
void handleMathFuncF(size_t funcCode, float op1, size_t op1Int,                        										float computedRes, size_t insIndex, size_t newRegIdx);
void handleMathFuncD(size_t funcCode, double op1, size_t op1Int,                       										double computedRes, size_t insIndex, size_t newRegIdx);
void handleMathFunc3Args(size_t funcCode, double op1,                                                                         size_t op1Int, double op2, 
                                          size_t op2Int, double op3, 
                                            size_t op3Int, double computedRes,
                                              size_t insIndex, size_t newRegIdx);
size_t getRegRes(size_t insIndex);
int handleCmp(mpfr_t *op1, mpfr_t *op2);
void handleSelect(void *Addr, size_t insIndex, double op);
void fpSanLoadFromShadowMemF(void *Addr, size_t insIndex, float opf,                                                     bool castFlag, size_t newRegIdx);
void fpSanLoadFromShadowMemD(void *Addr, size_t insIndex, double opd,                           													bool castFlag, size_t newRegIdx);
void checkBranch(double op1, size_t op1Int, double op2, size_t op2Int,                          			size_t fcmpFlag, bool computedRes, size_t insIndex, size_t lineNo);
void setRealConstantD(size_t toAddrInt, double value);
void setRealConstantF(size_t toAddrInt, float value);
void setRealTemp(size_t toAddrInt, size_t fromAddrInt, double op);
void handleCalloc(size_t toAddrInt, size_t size1, size_t size2);
void handleMalloc(size_t toAddrInt, size_t size);
void handleLLVMMemset(size_t toAddrInt, size_t val, size_t size);
void handleLLVMMemcpy(size_t toAddrInt, size_t fromAddrInt, size_t size);
int isNaN(mpfr_t real);

#endif
