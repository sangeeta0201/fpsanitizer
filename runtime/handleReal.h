#include <iostream>
#include <map>
#include <math.h>
#include <gmp.h>
#include <mpfr.h>
#include <vector>
#include <stack>

#define PRECISION 100

struct ErrorAggregate {
  double max_error;
  double total_error;
  long long int num_evals;
};

struct Real{
  mpfr_t mpfr_val;
};
double regIndex = 200; //assuming there are 200 constants in the program
std::map<size_t, struct ErrorAggregate*>errorMap;
//this will link ins index to index of result in shadow mem
std::stack<size_t> retTrack;
std::map<size_t, size_t>insMap;
std::map<size_t, size_t>returnMap;
std::map<std::map<size_t, size_t>, size_t> shadowFunArgMap; // thi will link function argument to shadowMap
std::map<size_t, struct Real*> shadowMap;

double getDouble(Real *real);
unsigned long ulpd(double x, double y);
void handleOp(size_t opCode, mpfr_t *res, mpfr_t *op1, mpfr_t *op2);
int isNaN(Real *real);
double updateError(Real *realVal, double computedVal, size_t insIndex);
void initializeErrorAggregate(ErrorAggregate *eagg);

extern "C" size_t getNewRegIndex();
extern "C" void* handleOp_1(size_t opCode, void *op1, void *op2);
extern "C" void* handleOp_2_f(size_t opCode, float op1, void *op2);
extern "C" void* handleOp_2_d(size_t opCode, double op1, void *op2);
extern "C" void* handleOp_3(size_t opCode, void *op1, float op2);
extern "C" void* handleOp_4_ff(size_t opCode, float op1, float op2);
extern "C" void* handleOp_4_fd(size_t opCode, float op1, double op2);
extern "C" void* handleOp_4_df(size_t opCode, double op1, float op2);
extern "C" void* handleOp_4_dd(size_t opCode, double op1, double op2);
extern "C" size_t handleOp_rd(size_t opCode, size_t regIndex, double op2);
extern "C" size_t setRealReg(size_t index, double value);
extern "C" void addFunArg(size_t argNo, void *funAddr, void *argAddr);
extern "C" void printErrorF(size_t result_A, float y);
extern "C" void printErrorD(size_t result_A, double y);
/*
When we compute any floating point operation on real, we save the result in shadow map with its value 
and return the address of this saved value to LLVM Pass. LLVM pass creates the mapping of real result and 
floating result. Wherever floating result is used, we use real result to compute the parallel operation in real.
Whenever store is called to save the value of temp into variable, we call setRealTemp. This function save the real operation 
result into variable.  
we create a ma
For every store, we call setRealConstant if  store is storing constant to some variable. 
We create shadow space for this variable, which is address of varaible and the real constant 
of this variable.
It 
*/
extern "C" void setRealTemp(void *toAddr, void *fromAddr);

/*
For every store, we call setRealConstant if  store is storing constant to some variable. 
We create shadow space for this variable, which is address of varaible and the real constant 
of this variable.
It 
*/
extern "C" void setRealConstant_float(void *Addr, float value);
extern "C" void setRealConstant_double(void *Addr, double value);
