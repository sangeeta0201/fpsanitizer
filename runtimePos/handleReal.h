#include <iostream>
#include <map>
#include <math.h>
#include <gmp.h>
#include <mpfr.h>
#include <vector>
#include <stack>
#include <list>
#include <assert.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <sys/mman.h>
#include "softposit.h"
#define PRECISION 1000
#define BUFLEN 10000000
#define MMAP_FLAGS (MAP_PRIVATE| MAP_ANONYMOUS| MAP_NORESERVE)
#define MAX_STACK_SIZE 1000000000
#define MAX_SIZE 100000000

struct ErrorAggregate {
  double max_error;
  double total_error;
  long long int num_evals;
};
struct BrError {
  bool corrRes;
  bool incorrRes;
	size_t lineNo;
  long long int num_evals;
};

struct Real{
  mpfr_t mpfr_val;
};

struct MyShadow{
	size_t key;
	struct Real * real;
};
size_t argCount = 0;
int frameIdx = 0;
int slotIdx = -1;
int returnIdx = -1;
size_t newRegIdx = 0;
bool recurFlag = false;
size_t mpfrClear = 0;
size_t mpfrInit = 0;
int varCount = 0;
int opCount = 0;
size_t funcCount = 0;
std::map<size_t, struct ErrorAggregate*>errorMap;
std::map<size_t, struct BrError*>errBrMap;
//this will link ins index to index of result in shadow mem
//std::list<struct MyShadow*> varTrack;
struct MyShadow *varTrack;
posit32_t *shadowStack;
posit32_t *shadowMap;
size_t *insMap;
size_t *frameCur; 
std::stack<size_t> retTrack;
std::map<size_t, size_t>funRetMap;

std::map<std::map<size_t, size_t>, size_t> shadowFunArgMap; // thi will link function argument to shadowMap

double getDouble(posit32_t val);
unsigned long ulpd(double x, double y);
void handleOp(size_t opCode, mpfr_t *res, mpfr_t *op1, mpfr_t *op2);
int isNaN(Real *real);
void initializeBrError(BrError *err);
double updateError(posit32_t val, double computedVal, size_t insIndex);
void updateBranchError(bool realRes, bool computedRes, size_t insIndex, size_t lineNo);
void initializeErrorAggregate(ErrorAggregate *eagg);
void printReal(Real *real);
/*
When we compute any floating point operation on real, we save the result in shadow map with its value 
and return the address of this saved value to LLVM Pass. LLVM pass creates the mapping of real result and 
floating result. Wherever floating result is used, we use real result to compute the parallel operation in real.
Whenever store is called to save the value of temp into variable, we call setRealTemp. This function save the real operation 
result into variable.  
we create a ma
For every store, we call setRealConstant if  store is storing constant to some variable. 
We create shadow space for this variable, which is address of varaible and the real constant 
of this variable.
It 
*/
//extern "C" void setRealTemp(void *toAddr, void *fromAddr);

/*
For every store, we call setRealConstant if  store is storing constant to some variable. 
We create shadow space for this variable, which is address of varaible and the real constant 
of this variable.
It 
*/
extern "C" void setRealConstant_float(void *Addr, float value);
extern "C" void setRealConstant_double(void *Addr, double value);
