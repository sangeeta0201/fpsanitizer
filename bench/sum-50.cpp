#include <stdio.h>
#include "handleReal.h"

int main() {
  volatile double x;
  for (x = 0.0; x < 10.0; x += 0.2);
  printf("%.20g\n", x);
  size_t result_A = (size_t) &x;
  printError(result_A, x);
}
