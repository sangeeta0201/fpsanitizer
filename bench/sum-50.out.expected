(output
  (function "main")
  (filename "sum-50.c")
  (line-num 6)
  (instr-addr 1087D9)
  (avg-error 2.000000)
  (max-error 2.000000)
  (num-calls 1)
  (influences
    (
    )
  )
)
(compare
  (function "main")
  (filename "sum-50.c")
  (line-num 5)
  (instr-addr 1087AE)
  (percent-wrong 1)
  (num-wrong 1)
  (num-calls 52)
  (influences
    (
    )
  )
)

